src/Test.UnitTests.Sarif.Driver/Sdk/AnalyzeCommandBaseTests.cs

```diff
@@ -1258,21 +1258,93 @@ private void AnalyzeScenarios(int[] scenarios)
+            const int expectedNumberOfArtifacts = 2;
+            const int expectedNumberOfResultsWithErrors = 1;
+            const int expectedNumberOfResultsWithWarnings = 1;
+            var files = new List<string>
+            {
+                $@"{Environment.CurrentDirectory}\Error.dll",
+                $@"{Environment.CurrentDirectory}\Warning.dll",
+                $@"{Environment.CurrentDirectory}\Note.dll",
+                $@"{Environment.CurrentDirectory}\Pass.dll",
+                $@"{Environment.CurrentDirectory}\NotApplicable.exe",
+                $@"{Environment.CurrentDirectory}\Informational.sys",
+                $@"{Environment.CurrentDirectory}\Open.cab",
+                $@"{Environment.CurrentDirectory}\Review.dll",
+                $@"{Environment.CurrentDirectory}\NoIssues.dll",
+            };
+
+            var testCase = new ResultsCachingTestCase
+            {
+                Files = files,
+                PersistLogFileToDisk = true,
+                FileSystem = CreateDefaultFileSystemForResultsCaching(files, generateSameInput: false)
+            };
+
+            var options = new TestAnalyzeOptions
+            {
+                TestRuleBehaviors = testCase.TestRuleBehaviors,
+                OutputFilePath = testCase.PersistLogFileToDisk ? Guid.NewGuid().ToString() : null,
+                TargetFileSpecifiers = new string[] { Guid.NewGuid().ToString() },
+                Kind = new List<ResultKind> { ResultKind.Fail },
+                Level = new List<FailureLevel> { FailureLevel.Warning, FailureLevel.Error },
+                DataToInsert = new OptionallyEmittedData[] { OptionallyEmittedData.Hashes },
+            };
+
+            Run run = RunAnalyzeCommand(options, testCase, multithreaded: true);
+
+            // Hashes is enabled and we should expect to see two artifacts because we have:
+            // one result with Error level and one result with Warning level.
+            run.Artifacts.Should().HaveCount(expectedNumberOfArtifacts);
+            run.Results.Count(r => r.Level == FailureLevel.Error).Should().Be(expectedNumberOfResultsWithErrors);
+            run.Results.Count(r => r.Level == FailureLevel.Warning).Should().Be(expectedNumberOfResultsWithWarnings);
+
+            options.DataToInsert = new List<OptionallyEmittedData>();
+
+            run = RunAnalyzeCommand(options, testCase, multithreaded: true);
+
+            // Hashes is disabled so no artifacts are expected.
+            run.Artifacts.Should().BeNull();
+            run.Results.Should().HaveCount(expectedNumberOfArtifacts);
+            run.Results.Count(r => r.Level == FailureLevel.Error).Should().Be(expectedNumberOfResultsWithErrors);
+            run.Results.Count(r => r.Level == FailureLevel.Warning).Should().Be(expectedNumberOfResultsWithWarnings);
+        }
+
+        [Fact]
+        public void AnalyzeCommmandBase_SingleThreaded_ShouldOnlyLogArtifactsWhenHashesIsEnabled()
```

Comment:
>AnalyzeCommmandBase_SingleThreaded_ShouldOnlyLogArtifactsWhenHashesIsEnabled [](http://example.com/codeflow?start=20&length=76)

This test will guarantee that we are always producing the correct number of artifacts if hashes is enabled for the single threaded analysis.

If not enabled, the artifacts should be null.